### ISSUE TITLE
Summit topic: Promises

### ISSUE THREAD

**Issue by MayaLekova** [nodejs/summit#86]
Creating this issue in attempt to gather links and offline discussion relevant to what's happening around the Promises topic.

### Motivation:
Make promises and other asynchronous primitives competitive performance-wise to traditional callback-oriented programming and easier to trace and debug.
* Why do we want to use promises and async/await?
* Recent optimization effort done in V8 and follow-up issues.
* Gathering real-world usage scenarios.
* [V8 API for promises](https://github.com/nodejs/summit/issues/60#issuecomment-388754878)
* [Improving debugging for promises](https://github.com/nodejs/help/issues/744) in Node.js
* Other related sub-topics - [async hooks](https://nodejs.org/api/async_hooks.html), [zones](https://github.com/domenic/zones).

### Events:
* 9:30 - 10:00 (main space) Talk: Holding on to your Performance Promises @bmeurer @MayaLekova
* 13:30 - 15:00 (meeting room) Workshop: Promises	@bmeurer @benjamingr

### Resources:
// Please add whatever you feel contributes to the discussion
// Talk slides to be linked here

**Comment by benjamingr** [comment]
Hey @MayaLekova thank you for opening this :) Looking forward to your workshop about promise performance!

A couple of us have written user use-cases with promises in: https://github.com/nodejs/promise-use-cases and what userland libraries currently do better.

Namely, after doing a community survey, going over code on GitHub and asking around and also based on my own intuition being involved with bluebird and other promise libraries for a while and answering 700 SO questions - what I saw most and intend to discuss [written here](https://github.com/nodejs/promise-use-cases/tree/master/use-cases) is:

 - Debugging issues like [issues with `unhandledRejection` hooks and debugging](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/unhandled-rejections/unhandled-rejections-1/unhandled-rejections-1.md) or [people using `.catch(error => console.log("error")` and why they do that](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/unhandled-rejections/unhandled-rejections-3/unhandled-rejections-3.md)
 - Several user errors people run into a lot like [forgetting to return a promise](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/1/1.md), [forgetting to `await`](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/2/2.md), [`resolve`ing/`reject`ing multiple times](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/3/3.md), have [never resolving promises](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/4/4.md), [using promises in a `forEach`](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/user-expectations/4/expectations-4.md) or getting [deadlocks](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/5/5.md)
 - Problematic usage patterns like [explicit construction](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/user-expectations/1/expectations-1.md), [having a hard time creating promises correctly](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/user-expectations/2/expectations-2.md),
 - People [ask how promises work in Node.js](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/user-expectations/3/expectations-3.md) and we don't have material ready about it.
 - It is [currently impossible/very hard to use testing libraries like lolex with promises](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/testing/testing-1/testing-1.md), or [mock race conditions](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/testing/testing-2/testing-2.md).
 - People are having a [hard time with stack traces when testing](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/stack-traces/1/stack-traces-1.md), [or when debugging production issues](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/stack-traces/2/stack-traces-2.md) and [postmortem is an open issue](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/stack-traces/3/stack-traces-3.md).
 - People [still use userland libraries like bluebird](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/performance/1/performance-1.md) and [maybe avoid async iterators](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/performance/2/performance-2.md).
 - There are a lot of things people ask for like [hooks for context](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/forgotten-promises/1/forgotten-1.md), the [ability to build cancellation](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/extras/1/extras-1.md) (also [2](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/extras/3/extras-3.md)) or [use it in a standardized way](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/extras/2/extras-2.md) 

I don't want to suggest any actions but rather bring the issues up and discuss them and see what the people attending the workshop think about any of these. My personal feeling is that people I talked to felt the most strongly about testing async functions being hard and stack traces.

**Comment by bmeurer** [comment]
Woah. Thanks @MayaLekova for kicking this off, and thanks @benjamingr for this nicely structured list of use cases! Looks like there's a lot to discuss for us. 👍

**Comment by bmeurer** [comment]
@benjamingr That matches my expectations. We hope to develop a better understanding of the needs of developers specifically on the Node side, so that we can drive future work on V8 in the right direction. 👍

**Comment by MayaLekova** [comment]
Being pretty new to this topic, my expectations are also around listening to what's important for the community, so we can better focus our efforts in V8.
Thank you for structuring this use cases so well!

**Comment by benjamingr** [comment]
Hey, really nice talk @MayaLekova and @bmeurer ! 

Some discussion points:
 - We've had the same issues with monitoring and setScheduler in bluebird where the scheduling bypass interfered with change detection people used in Angular
 - I'd love to see some performance measures of coroutines as `async`/`await` vs. `async`/`await` and how important the actual ability to hook on microtick scheduling semantics.
 - I'd love to talk about _why_ the spec works this way, the whole "delaying a microtick" issue was due to a real concern (avoiding zalgo) but there is no reason I can think of that the spec couldn't/shouldn't allow elision of promises in that scenario - have you considered bringing it up with TC39?

**Comment by JiaLiPassion** [comment]
@benjamingr ,for the `bluebird` angular issue, it has been resolved in `zone.js`, the document can be found here, https://github.com/angular/zone.js/blob/master/NON-STANDARD-APIS.md (search bluebird).

@bmeurer  And I want to know current `ES2017 native async/await` use `native promise` which can not intercepted, in `nodejs` , we can use `asynchooks` and `PromiseHooks`, but in browser, currently we can not do anything about it. Is `PromiseHooks` or `AsyncHooks` javascript API for in the future plan? 

Thank you.

**Comment by benjamingr** [comment]
@JiaLiPassion Zones are a no-go for Node.js at the moment due to the error handling issues, which is sort of a shame since a lot of us are using promises now (and the issue is basically about nested callbacks).

**Comment by bmeurer** [comment]
@JiaLiPassion that's something to discuss with @domenic and @benlesh

**Comment by benjamingr** [comment]
Interesting discussion in the Diagnostics session about zones and async_hooks - I think the current PoV is that Node.js needs to figure out async_hooks error handling before we unblock zones.

**Comment by JiaLiPassion** [comment]
@benjamingr , @bmeurer , got it, thank you!

**Comment by JiaLiPassion** [comment]
@benjamingr , yeah, currently the issue is in `angular` and `rxjs test`, in the near future , if user want to compile `typescript` code` into `ES2017 target` with `async/await`, `zone.js` can't control the `native promise`. So user will need to handle `change detection` and `async test` themselves.

**Comment by benjamingr** [comment]
@JiaLiPassion IIRC Ben Lesh said "IE11 is saving the internet".

This is something I plan to bring up in the second session - but TC39 is going to need to solve this for async functions and not Node.js. Feel free to open an issue in https://github.com/nodejs/promise-use-cases

**Comment by JiaLiPassion** [comment]
@benjamingr , got it, I will!

**Comment by mcollina** [comment]
You can join remotely at https://zoom.us/j/903800260.

**Comment by benjamingr** [comment]
@MayaLekova can you post the meeting notes :)?

**Comment by MayaLekova** [comment]
Here are the [meeting notes](https://docs.google.com/document/d/1xKfCG0EsNNES4X9M4xOj-DxCOVMunUXNK3zsWFPJCfc/edit?usp=sharing)
The `#number` in the notes reference slide titles from @benjamingr slides (if you can share them as well would be awesome!).

**Comment by benjamingr** [comment]
https://docs.google.com/presentation/d/16WYzlvvDm85IruD3jkH2nTwTkTUa6YSPtDF0VQ4g4Hk/edit?usp=sharing

**Comment by benjamingr** [comment]
PTAL https://github.com/petkaantonov/bluebird/pull/1530/files

**Comment by benjamingr** [comment]
@bmeurer @MayaLekova two questions:

 - Can we close this issue? (Or do you want to do another session today?)
 - Can/should we schedule a follow-up meeting online on this in a month (two?)?

**Comment by bmeurer** [comment]
I have to leave early today, but I'm happy to talk about promises in the morning. I'd say we can close this.

**Comment by benjamingr** [comment]
https://doodle.com/poll/3wdzdfzikmcvndee

**Comment by benjamingr** [comment]
Thanks for participating everything I learned a lot and I think a lot of good progress was made. Please consider filling the doodle above if you haven't already. Also pinging @BridgeAR @addaleax @littledan in case either of them would like to also participate in the follow up discussion.

**Comment by littledan** [comment]
@benjamingr Apologies for my delay; I've filled out the doodle.

**Comment by BridgeAR** [comment]
The same here.

**Comment by benjamingr** [comment]
Great, thanks for the update - going to wait for another day or two to see if anyone else wants to participate -the most likely candidate is Thu, Jul 12 @ 3:30 PM UTC unless we can figure out something that also works for @ljharb
